### DISCUSSION TITLE
Accessing array variables on destination side using mirth

### DISCUSSION THREAD

**Discussion by asujata-1234** [nextgenhealthcare/connect#4964] [Q&A]
Hello
I am trying to access array variables 

Source transformer script to create an array of all the billing provider Id's in an 837P claim that has multiple CLM segments

`var BillingProviderId = new Array();
var counter = 0;
for (var i = 0; i < getArrayOrXmlLength(msg['NM1']); i++) {
	msgNM1 = (msg['NM1'][i]['NM1.01']['NM1.01.1'].toString());
	if (msgNM1 == '85') {
		BillingProviderId[counter] = (msg['NM1'][i]['NM1.09']['NM1.09.1'].toString());
		counter += 1;
	channelMap.put("BillingProviderId",BillingProviderId);
}
}`

On destination side I am trying to access this array variable  BillingProviderId and write to DB. once record per array element. I was able to write to DB entire array.

My destination side code is
`var dbConn;
var result;
try {
	dbConn = DatabaseConnectionFactory.createDatabaseConnection('net.sourceforge.jtds.jdbc.Driver','jdbc:sqlserver://172.00.00.65:1433;instanceName=DEVSQL;databaseName=GeneralHospital;integratedSecurity=true);
	var params = new java.util.ArrayList();
	var BillingProviderId = new Packages.java.util.ArrayList();
	var BillingProviderId = channelMap.get("BillingProviderId");
	
	params.add($('BillingProviderId'));
	//params.add($('BillingProviderId')[i]);
	params.add($('BillingProviderOrganizationName'));
	params.add($('ClaimNumber'));
	params.add($('MedicalRecordNumber'));
	params.add($('SubscriberId'));
	params.add($('SubscriberLastName'));
	params.add($('SubscriberFirstName'));
	params.add($('SubscriberDateOfBirth'));
	params.add($('SubscriberGender'));
	params.add($('PatientIsSubscriber'));
	params.add($('Payer'));
	params.add($('PatientId'));
	params.add($('PatientLastName'));
	params.add($('PatientFirstName'));
	params.add($('PatientDateOfBirth'));
	params.add($('PatientGender'));
	params.add($('TotalCharges'));
	params.add((2));
	params.add($('ClaimType'));
	params.add($('AdmitDate'));
	params.add($('DischargeDate'));
var expression = "INSERT INTO stage.StagedClaim (BillingProviderId,BillingProviderOrganizationName, ClaimNumber, MedicalRecordNumber, SubscriberId, SubscriberLastName, SubscriberFirstName, SubscriberDateOfBirth, SubscriberGender, PatientIsSubscriber, Payer, PatientId,PatientLastName,PatientFirstName,PatientDateOfBirth,PatientGender,TotalCharges,ClaimStatusId,ClaimType,AdmitDate,DischargeDate) VALUES (?,?, ?, ?, ?, ?, ?, ?, ?, ?, ?,?,?,?,?,?,?,?,?,?,?)";
	
var result = dbConn.executeUpdate(expression, params);
} finally {
	if (dbConn) { 
		dbConn.close();
	}
}`
above code write to DB whole array in one row in DB at this time other variables are not arrays.
I did try to iterate over the array BillingProviderId and write insert query inside for loop and outside for loop as well but I was not successful. Any suggestions will be helpful

**Comment by rjkroll** [comment]
This is a very basic, basic example, but hopefully it will lead you in the right direction.    In your Source transformer, you need to do something like this so that its available in the Destination.   Use the last line of code in your Destination to pull in the variable(s) from the Source:

```
//declare arraylist or any other object.
var arrayNam = new Packages.java.util.ArrayList();

// do with it what you want
arrayNam.append("text");

//put it in a channelMap
channelMap.put("ArrayNam",arrayNam);


//at the destination transformer retrieve the arrayList:
var arrayNam = channelMap.get("ArrayNam");
```

**Reply by pacmano1** [reply]
You might also consider that your DB supports arrays natively.   So iteration might not be needed - also you can "mass edit" XML or JSON arrays as needed.

To the OP - take a look at the code highlighting suggestions at https://github.com/nextgenhealthcare/connect/discussions/4824 and https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

Your code above highlighted:
```JS
var BillingProviderId = new Array();
var counter = 0;
for(var i = 0; i < getArrayOrXmlLength(msg['NM1']); i++) {
    msgNM1 = (msg['NM1'][i]['NM1.01']['NM1.01.1'].toString());
    if(msgNM1 == '85') {
        BillingProviderId[counter] = (msg['NM1'][i]['NM1.09']['NM1.09.1'].toString());
        counter += 1;
        channelMap.put("BillingProviderId", BillingProviderId);
    }
}
```
And:

```JS
var dbConn;
var result;
try {
    dbConn = DatabaseConnectionFactory.createDatabaseConnection('net.sourceforge.jtds.jdbc.Driver', 'jdbc:sqlserver://172.00.00.65:1433;instanceName=DEVSQL;databaseName=GeneralHospital;integratedSecurity=true);
        var params = new java.util.ArrayList();
        var BillingProviderId = new Packages.java.util.ArrayList();
        var BillingProviderId = channelMap.get("BillingProviderId");

        params.add($('BillingProviderId'));
        //params.add($('BillingProviderId')[i]);
        params.add($('BillingProviderOrganizationName')); params.add($('ClaimNumber')); params.add($('MedicalRecordNumber')); params.add($('SubscriberId')); params.add($('SubscriberLastName')); params.add($('SubscriberFirstName')); params.add($('SubscriberDateOfBirth')); params.add($('SubscriberGender')); params.add($('PatientIsSubscriber')); params.add($('Payer')); params.add($('PatientId')); params.add($('PatientLastName')); params.add($('PatientFirstName')); params.add($('PatientDateOfBirth')); params.add($('PatientGender')); params.add($('TotalCharges')); params.add((2)); params.add($('ClaimType')); params.add($('AdmitDate')); params.add($('DischargeDate'));
        var expression = "INSERT INTO stage.StagedClaim (BillingProviderId,BillingProviderOrganizationName, ClaimNumber, MedicalRecordNumber, SubscriberId, SubscriberLastName, SubscriberFirstName, SubscriberDateOfBirth, SubscriberGender, PatientIsSubscriber, Payer, PatientId,PatientLastName,PatientFirstName,PatientDateOfBirth,PatientGender,TotalCharges,ClaimStatusId,ClaimType,AdmitDate,DischargeDate) VALUES (?,?, ?, ?, ?, ?, ?, ?, ?, ?, ?,?,?,?,?,?,?,?,?,?,?)";

        var result = dbConn.executeUpdate(expression, params);
    }
    finally {
        if(dbConn) {
            dbConn.close();
        }
    }
```

**Reply by asujata-1234** [reply]
Thank you for your responses. I was able to retrieve the arrayList at the destination. But wile inserting records all the elements of an array are inserted in one single row.
![image](https://user-images.githubusercontent.com/19535802/150022532-1e04c050-27eb-4a5e-8c25-2156924275de.png)

**Reply by pacmano1** [reply]
You will have to iterate over the array and process it one item at a time if you decide not to use native SQL array functions for the insert.

The code you posted above is not iterating.

**Reply by asujata-1234** [reply]
I used JSON.Parse and iterated over an ArrayList. Now the records are inserted in DB when there are multiple claims in a single file.

**Reply by asujata-1234** [reply]
I am trying to write all the subscriber last names when there are multiple claims in a file and claim has two payers primary and secondary. my code is reading all subscriber last names and writing to an array. So for a claim with payers has 2 subscriber last names. Since I am iterating over an array on claim number next claim is getting the wrong subscriber.. So far I tried to write a condition to check on primary claim but does not seem to be working. Any help is appreciated.

```
var SubscriberLastName = new Array();

//var BillingProviderOrganizationName = new Packages.java.util.ArrayList();
var counter = 0;
for (var i = 0; i < getArrayOrXmlLength(msg['SBR']); i++) {
	msgsbr03 = (msg['SBR'][i]['SBR.01']['SBR.01.1'].toString());
	if (msgsbr03 =='P') {
		for (var i = 0; i < getArrayOrXmlLength(msg['NM1']); i++) {
			//msgsbr05 = (msg['SBR'][i]['SBR.01']['SBR.01.1'].toString());
			msgNM3 = (msg['NM1'][i]['NM1.01']['NM1.01.1'].toString());
			if (msgNM3 =='IL') {
				SubscriberLastName[counter] = (msg['NM1'][i]['NM1.03']['NM1.03.1'].toString());
				counter += 1;
			channelMap.put("SubscriberLastName",SubscriberLastName);
}
}
	}}
```

**Reply by pacmano1** [reply]
Again, **please use code highlighting.**

```JS
var SubscriberLastName = new Array();

//var BillingProviderOrganizationName = new Packages.java.util.ArrayList();
var counter = 0;
for(var i = 0; i < getArrayOrXmlLength(msg['SBR']); i++) {
    msgsbr03 = (msg['SBR'][i]['SBR.01']['SBR.01.1'].toString());
    if(msgsbr03 == 'P') {
        for(var i = 0; i < getArrayOrXmlLength(msg['NM1']); i++) {
            //msgsbr05 = (msg['SBR'][i]['SBR.01']['SBR.01.1'].toString());
            msgNM3 = (msg['NM1'][i]['NM1.01']['NM1.01.1'].toString());
            if(msgNM3 == 'IL') {
                SubscriberLastName[counter] = (msg['NM1'][i]['NM1.03']['NM1.03.1'].toString());
                counter += 1;
                channelMap.put("SubscriberLastName", SubscriberLastName);
            }
        }
    }
}
```
Rather than read through your code, post a sample message and the objective.

**Comment by asujata-1234** [comment]
[testfile.txt](https://github.com/nextgenhealthcare/connect/files/7938487/testfile.txt)
This is the sample message. This claim has two payers primary and secondary with two same subscribers. I am trying to read first subscriber lastname, firstname subscriberid for a primary payer and trying to write to an array. My code is writing both subscriber details and writing to an array.

**Reply by pacmano1** [reply]
I don't do x12, but I think you need to stop "collecting' segments in your inside loop.

Using parts of your code that I could:

```js
for(var i = 0; i < msg['SBR'].length(); i++) {
    if(msg['SBR'][i]['SBR.01']['SBR.01.1'].toString() == 'P') {
        for each(nm1 in getSegmentsAfter(msg, msg['SBR'][i], 'NM1', false, 'SBR')) {
            if(nm1['NM1.01']['NM1.01.1'].toString() == 'IL') {
                $c('subscriber_ln', nm1['NM1.03']['NM1.03.1'].toString())
                var found = true;
            }
        }
    }
    if(found == true) break;
}
```
See https://github.com/nextgenhealthcare/connect-examples/tree/master/Code%20Templates/Get%20Segments%20After%20a%20Particular%20Segment

You may find use with https://github.com/nextgenhealthcare/connect-examples/tree/master/Code%20Templates/Filter%20XMLLists

**Comment by asujata-1234** [comment]
```javascript
#This code works if I have just one claim in a message and file is not getting picked up and remains in received status for long time when there are multiple claims.
var SubscriberLastName = new Array();
var counter = 0;
for(var i = 0; i < msg['SBR'].length(); i++) {
    if(msg['SBR'][i]['SBR.01']['SBR.01.1'].toString() == 'P') {
        for each(SBR in getSegmentsAfter(msg, msg['SBR'][0], 'NM1', false,msg['SBR'][1])) {
            if(msg['NM1'][i]['NM1.01']['NM1.01.1'].toString() == 'IL') {
            	SubscriberLastName[counter] = (msg['NM1'][i]['NM1.03']['NM1.03.1'].toString());
			counter += 1;
		  channelMap.put("SubscriberLastName",SubscriberLastName);
               var found = true;
            }
        }
    }
    if(found == true) break;
}
```


I commented out the code var found and dropped same message with multiple claims. same last name keeps on storing 6 times. I know the reason why but so far these are the things I tried.
```javascript
var SubscriberLastName = new Array();
var counter = 0;
for(var i = 0; i < msg['SBR'].length(); i++) {
    if(msg['SBR'][i]['SBR.01']['SBR.01.1'].toString() == 'P') {
        for each(SBR in getSegmentsAfter(msg, msg['SBR'][0], 'NM1', false,msg['SBR'][1])) {
            if(msg['NM1'][i]['NM1.01']['NM1.01.1'].toString() == 'IL') {
            	SubscriberLastName[counter] = (msg['NM1'][i]['NM1.03']['NM1.03.1'].toString());
			counter += 1;
		  channelMap.put("SubscriberLastName",SubscriberLastName);
              // var found = true;
            }
        }
    }
   // if(found == true) break;
```

**Reply by pacmano1** [reply]
`channelMap.put("SubscriberLastName",SubscriberLastName);`

should be the below if you want to see the channelMap var in the message viewer for each iteration.

`channelMap.put("SubscriberLastName_" + i,SubscriberLastName); `

**Reply by asujata-1234** [reply]
I am sorry, I am having so much trouble with this. I requested to restart the server. Even after deploying channels keeps on going to stopped status. I will be right back. I tried so many variations so far. I also tried to child segment length
```javascript
var SubscriberLastName = new Array();

//var BillingProviderOrganizationName = new Packages.java.util.ArrayList();
var counter = 0;
for (var i = 0; i < getArrayOrXmlLength(msg['NM1']); i++) {
	var sbrGroup = getSegmentsAfter(msg,msg.SBR[0],'NM1',true);
	msgNM3 = (msg['NM1'][i]['NM1.01']['NM1.01.1'].toString());
	if (sbrGroup.length > 0){
     	msg.children()[sbrGroup[sbrGroup.length-1].childIndex()];
     	if (msgNM3 == 'IL') {
				SubscriberLastName[counter] = (msg['NM1'][i]['NM1.03']['NM1.03.1'].toString());
				counter += 1;
			channelMap.put("SubscriberLastName",SubscriberLastName);}
	else {
	break;
	}
	}
}
```

**Reply by asujata-1234** [reply]
It is throwing an error .
Transformer error
ERROR MESSAGE: Error evaluating transformer
com.mirth.connect.server.MirthJavascriptTransformerException: 
CHANNEL:	WTN-Healthcare-837P-01-25
CONNECTOR:	sourceConnector
SCRIPT SOURCE:	TRANSFORMER
SOURCE CODE:	
250:     if(msg['SBR'][i]['SBR.01']['SBR.01.1'].toString() == 'P') {
251:         for each(SBR in getSegmentsAfter(msg, msg['SBR'][0], 'NM1', false,msg['SBR'][1])) {
252:             if(msg['NM1'][i]['NM1.01']['NM1.01.1'].toString() == 'IL') {
253:             	SubscriberLastName[counter] = (msg['NM1'][i]['NM1.03']['NM1.03.1'].toString());
254: 			counter += 1;
255: 		  channelMap.put("SubscriberLastName",+ i,SubscriberLastName);
256: 		//  channelMap.put("SubscriberLastName_" + i,SubscriberLastName);
257:              //  var found = true;
258:             }
259:         }
LINE NUMBER:	255
DETAILS:	Can't find method com.mirth.connect.server.userutil.ChannelMap.put(string,number,object).

**Reply by pacmano1** [reply]
You have a comma in yours:

`channelMap.put("SubscriberLastName",+ i,SubscriberLastName);`

I think in your code you want:

`channelMap.put("SubscriberLastName" + i,msg['NM1'][i]['NM1.03']['NM1.03.1'].toString())`

**Reply by asujata-1234** [reply]
I think I solved the problem. Now subscriber and patient names are inserting in DB when there are multiple claims. On to next challenge

```javascript
var SubscriberLastName = new Array();
var counter = 0;
for(var i = 0; i < msg['SBR'].length(); i++) {
   // if(msg['SBR'][i]['SBR.01']['SBR.01.1'].toString() == 'P') {
        for each((msg['NM1'][i]) in getSegmentsAfter(msg, msg['SBR'][i], 'NM1', true)) {
            if(msg['NM1'][i]['NM1.01']['NM1.01.1'].toString() == 'IL') {
            	SubscriberLastName[counter] = (msg['NM1'][i]['NM1.03']['NM1.03.1'].toString());
			counter += 1;
		  channelMap.put("SubscriberLastName",SubscriberLastName);
		//  channelMap.put("SubscriberLastName_" + i,SubscriberLastName);
              // var found = true;
            }
        }
   // }
 //  if(found == true) break;
}

```

**Comment by asujata-1234** [comment]


**Reply by pacmano1** [reply]
Well maybe:

```JS
for(var i = 0; i < msg['SBR'].length(); i++) {
    var j = 0
    if(msg['SBR'][i]['SBR.01']['SBR.01.1'].toString() == 'P') {
        for each(nm1 in getSegmentsAfter(msg, msg['SBR'][i], 'NM1', false, 'SBR')) {
            if(nm1['NM1.01']['NM1.01.1'].toString() == 'IL') {
                $c('subscriber_ln' + i + "_" + j, nm1['NM1.03']['NM1.03.1'].toString());
                j++
                                            
            }
        }
    }
}
```

![image](https://user-images.githubusercontent.com/44065187/152055071-02aca8cd-1575-4fd0-92bc-24ff441e623a.png)

**Reply by asujata-1234** [reply]
I can not predict how many claims will be in each file while accessing at the destination while inserting into DB will cause problem.
I posted above that code worked with some limitations when I came across tertiary subscriber.

**Comment by asujata-1234** [comment]
![image](https://user-images.githubusercontent.com/19535802/153070308-4039e7e1-ab00-4db7-beaf-58939d85c38a.png)
Below code solved the issue with above source transformation design. 


```javascript
var SubscriberFirstName = new Array();

var counter = 0;
for(var i = 0; i < msg['SBR'].length(); i++) {
//	if(msg['SBR'][i]['SBR.01']['SBR.01.1'].toString() == 'P') {
		
		for each(nm1 in getSegmentsAfter(msg, msg['SBR'][i], 'NM1', true)) {
		
			msgNM2 = (nm1['NM1.01']['NM1.01.1'].toString());			
			if (msgNM2 == 'IL') {
				SubscriberFirstName[counter] = (nm1['NM1.04']['NM1.04.1'].toString());
				counter += 1;
			channelMap.put("SubscriberFirstName",SubscriberFirstName);
}
}
}
'''

**Reply by asujata-1234** [reply]
@pacmano1 

Thank you so much for your response. I am trying to transform patient info when subscriber is not the patient along with some claims when subscriber is the patient. I came across another challenge when subscriber is not the patient or subscriber has dependent. So far I tried below code and few other variations. This code gives me all the subscribers for all claims and patient with qualifier QC. I do not need subscriber name KENDALL when claim has dependent "LAZARUS". I will be posting sample message. I am scrubbing file.

["MARY","EMORY","HUBERT","SHAWAN","TIERA","KENDELL","LAZARUS","GUY","LAUREN","REX","KIMBERLY","JOHN","AMANDA","CHHAY","JAMES","DAMONA"]
Processed sample file
["TFirst1","TFirst2","TFirst3","TFirst4","TFirst5","TFirst6","QC_TFirst6","TFirst7","TFirst8","TFirst9","TFirst10","TFirst11","TFirst12","TFirst13","TFirst14","TFirst15"]

Expected results

["TFirst1","TFirst2","TFirst3","TFirst4","TFirst5","QC_TFirst6","TFirst7","TFirst8","TFirst9","TFirst10","TFirst11","TFirst12","TFirst13","TFirst14","TFirst15"]

```javascript
var PatientFirstName = new Array();
var counter = 0;

for(var i = 0; i < msg['SBR'].length(); i++) {

		
		
				for each(nm1 in getSegmentsAfter(msg, msg['SBR'][i], 'NM1', true)) {
					//msgNM2 = (nm1['NM1.01']['NM1.01.1'].toString());

							if (nm1['NM1.01']['NM1.01.1'].toString() == 'QC') {
								PatientFirstName[counter] = (nm1['NM1.04']['NM1.04.1'].toString());
								counter += 1;
							channelMap.put("PatientFirstName",PatientFirstName);}
								
							else if (nm1['NM1.01']['NM1.01.1'].toString() == 'IL') {
								
								
								PatientFirstName[counter] = (nm1['NM1.04']['NM1.04.1'].toString());
								counter += 1;
							channelMap.put("PatientFirstName",PatientFirstName);
							logger.info("PatientFirstName is=: " + PatientFirstName); 
				}}}
```

**Comment by asujata-1234** [comment]


**Reply by pacmano1** [reply]
I will defer to others here to help you out.  Best wishes with your project, I just don't have the time at the moment to deep dive on your code and use case.
